core/src/main/java/org/frankframework/pipes/EscapePipe.java

```diff
@@ -72,42 +72,52 @@ public PipeRunResult doPipe(Message message, PipeLineSession session) throws Pip
 			throw new PipeRunException(this, "cannot open stream", e);
 		}
 
-		String substring = null;
-		String result = input;
-		int i = -1;
-		int j = -1;
-		log.debug("substringStart [{}] substringEnd [{}] input [{}]", substringStart, substringEnd, input);
-		if (substringStart != null && substringEnd != null) {
-			i = input.indexOf(substringStart);
-			if (i != -1) {
-				j = input.indexOf(substringEnd, i);
-				if (j != -1) {
-					substring = input.substring(i + substringStart.length(), j);
-					substring = handle(substring);
-					result = input.substring(0, i + substringStart.length()) + substring + input.substring(j);
+		String result = handleSubstrings(input);
+		return new PipeRunResult(getSuccessForward(), result);
+	}
+
+	private String handleSubstrings(String input) {
+		if (substringStart == null || substringEnd == null) {
+			return handle(input);
+		}
+
+		StringBuilder result = new StringBuilder(input);
```

Comment:
I suggested using a StringBuilder not to do in-place changes in the StringBuilder but to gradually build the final result there, appending each piece as you process it.
The current code is very inefficient, causing a lot of overhead as strings are moved around over and over again in memory.

You can look at the implementation of `XmlEncodingUtils.encodeChars()` for an example of what I mean.

Reply:
Oops, I've missed that part. I didn't check/see how you've handled the prior feedback from Tim. Sorry for that.

Reply:
Ideally, for best performance, you can also refactor the code in XmlEncodingUtils encodeChars() / decodeChars() and the cdata handling code a bit to add overloaded methods that have a StringBuilder parameter to which they append. That will save a lot of additional copying too.
But if you will find that too confusing I can pick up that final optimisation.